File: sdk/samples/03_types/c/strings.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * Strings Sample - Demonstrates DDS string types
 *
 * This sample shows how to work with string types:
 * - Unbounded strings (char* pointer)
 * - Bounded strings (sequence<char> with .data/.len)
 * - Wide strings (wchar_t* pointer)
 */

#include <stdio.h>
#include <string.h>
#include <wchar.h>
#include "generated/Strings.h"

int main(void) {
    printf("=== HDDS String Types Sample ===\n\n");

    /* Create a Strings instance - new API uses char* pointers */
    Strings original;
    memset(&original, 0, sizeof(original));

    /* unbounded_str: char* pointer, assign string literal for encode */
    original.unbounded_str = "This is an unbounded string (up to buffer limit)";

    /* bounded_str: sequence<char> with .data and .len */
    char bounded_buf[] = "Bounded to 256 chars";
    original.bounded_str.data = bounded_buf;
    original.bounded_str.len = (uint32_t)strlen(bounded_buf);

    /* wide_str: wchar_t* pointer */
    wchar_t wide_buf[] = L"Wide string with UTF-8: Hello World!";
    original.wide_str = wide_buf;

    printf("Original Strings:\n");
    printf("  unbounded_str: \"%s\"\n", original.unbounded_str);
    printf("  bounded_str:   \"%.*s\" (max 256 chars)\n",
           (int)original.bounded_str.len, original.bounded_str.data);
    printf("  wide_str:      \"%ls\"\n", original.wide_str);

    /* Serialize */
    uint8_t buffer[4096];
    int serialized_size = strings_encode_cdr2_le(&original, buffer, sizeof(buffer));
    if (serialized_size <= 0) {
        printf("[ERROR] Serialization failed! (%d)\n", serialized_size);
        return 1;
    }
    printf("\nSerialized size: %d bytes\n", serialized_size);

    /* Deserialize - must pre-allocate buffers for char* fields */
    Strings deserialized;
    memset(&deserialized, 0, sizeof(deserialized));

    char deser_unbounded[512] = {0};
    deserialized.unbounded_str = deser_unbounded;

    char deser_bounded[256] = {0};
    deserialized.bounded_str.data = deser_bounded;
    deserialized.bounded_str.len = 0;

    wchar_t deser_wide[512] = {0};
    deserialized.wide_str = deser_wide;

    int deser_size = strings_decode_cdr2_le(&deserialized, buffer, (size_t)serialized_size);
    if (deser_size < 0) {
        printf("[ERROR] Deserialization failed! (%d)\n", deser_size);
        return 1;
    }

    printf("\nDeserialized:\n");
    printf("  unbounded_str: \"%s\"\n", deserialized.unbounded_str);
    printf("  bounded_str:   \"%.*s\"\n",
           (int)deserialized.bounded_str.len, deserialized.bounded_str.data);
    printf("  wide_str:      \"%ls\"\n", deserialized.wide_str);

    /* Verify round-trip */
    if (strcmp(original.unbounded_str, deserialized.unbounded_str) == 0 &&
        original.bounded_str.len == deserialized.bounded_str.len &&
        memcmp(original.bounded_str.data, deserialized.bounded_str.data,
               original.bounded_str.len) == 0 &&
        wcscmp(original.wide_str, deserialized.wide_str) == 0) {
        printf("\n[OK] Round-trip serialization successful!\n");
    } else {
        printf("\n[ERROR] Round-trip verification failed!\n");
        return 1;
    }

    /* Test empty strings */
    printf("\n--- Empty String Test ---\n");
    Strings empty;
    memset(&empty, 0, sizeof(empty));
    empty.unbounded_str = "";
    empty.bounded_str.data = NULL;
    empty.bounded_str.len = 0;
    wchar_t empty_wide[] = L"";
    empty.wide_str = empty_wide;

    int empty_size = strings_encode_cdr2_le(&empty, buffer, sizeof(buffer));

    Strings empty_deser;
    memset(&empty_deser, 0, sizeof(empty_deser));
    char empty_unbounded_buf[64] = {0};
    empty_deser.unbounded_str = empty_unbounded_buf;
    empty_deser.bounded_str.data = NULL;
    empty_deser.bounded_str.len = 0;
    wchar_t empty_wide_buf[64] = {0};
    empty_deser.wide_str = empty_wide_buf;

    strings_decode_cdr2_le(&empty_deser, buffer, (size_t)empty_size);

    if (strlen(empty_deser.unbounded_str) == 0) {
        printf("[OK] Empty strings handled correctly\n");
    }

    /* Test different length strings */
    printf("\n--- Various Length Test ---\n");
    Strings varied;
    memset(&varied, 0, sizeof(varied));
    varied.unbounded_str = "Short";

    char varied_bounded[256];
    memset(varied_bounded, 'X', 200);
    varied_bounded[200] = '\0';
    varied.bounded_str.data = varied_bounded;
    varied.bounded_str.len = 200;

    wchar_t varied_wide[] = L"Medium length string here";
    varied.wide_str = varied_wide;

    int varied_size = strings_encode_cdr2_le(&varied, buffer, sizeof(buffer));

    Strings varied_deser;
    memset(&varied_deser, 0, sizeof(varied_deser));
    char varied_deser_unbounded[512] = {0};
    varied_deser.unbounded_str = varied_deser_unbounded;
    char varied_deser_bounded[256] = {0};
    varied_deser.bounded_str.data = varied_deser_bounded;
    varied_deser.bounded_str.len = 0;
    wchar_t varied_deser_wide[512] = {0};
    varied_deser.wide_str = varied_deser_wide;

    strings_decode_cdr2_le(&varied_deser, buffer, (size_t)varied_size);

    printf("String lengths:\n");
    printf("  unbounded_str: %zu chars\n", strlen(varied_deser.unbounded_str));
    printf("  bounded_str:   %u chars\n", varied_deser.bounded_str.len);
    printf("  wide_str:      %zu chars\n", wcslen(varied_deser.wide_str));

    if (varied.bounded_str.len == varied_deser.bounded_str.len) {
        printf("[OK] Various length strings handled correctly\n");
    }

    printf("\n=== Sample Complete ===\n");
    return 0;
}

File: sdk/samples/03_types/python/strings.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Strings Sample - Demonstrates DDS string types

This sample shows how to work with string types:
- Unbounded strings
- Bounded strings (with length limit)
- Wide strings (wstring)
"""

import sys
sys.path.insert(0, '.')

from generated.Strings import Strings


def main():
    print("=== HDDS String Types Sample ===\n")

    # Create a Strings instance
    original = Strings(
        unbounded_str="This is an unbounded string that can be any length",
        bounded_str="Bounded to 256 chars",
        wide_str="Wide string with UTF-8: Héllo Wörld! 你好世界 🌍",
    )

    print("Original Strings:")
    print(f'  unbounded_str: "{original.unbounded_str}"')
    print(f'  bounded_str:   "{original.bounded_str}" (max 256 chars)')
    print(f'  wide_str:      "{original.wide_str}"')

    # Serialize
    data = original.encode_cdr2_le()
    print(f"\nSerialized size: {len(data)} bytes")

    # Deserialize
    deserialized, _ = Strings.decode_cdr2_le(data)
    print("\nDeserialized:")
    print(f'  unbounded_str: "{deserialized.unbounded_str}"')
    print(f'  bounded_str:   "{deserialized.bounded_str}"')
    print(f'  wide_str:      "{deserialized.wide_str}"')

    # Verify round-trip
    if original == deserialized:
        print("\n[OK] Round-trip serialization successful!")
    else:
        print("\n[ERROR] Round-trip verification failed!")
        return 1

    # Test empty strings
    print("\n--- Empty String Test ---")
    empty = Strings(unbounded_str="", bounded_str="", wide_str="")
    empty_data = empty.encode_cdr2_le()
    empty_deser, _ = Strings.decode_cdr2_le(empty_data)

    if empty == empty_deser:
        print("[OK] Empty strings handled correctly")

    # Test UTF-8 special characters
    print("\n--- UTF-8 Special Characters Test ---")
    utf8_test = Strings(
        unbounded_str="ASCII only: Hello World!",
        bounded_str="Latin-1: café résumé naïve",
        wide_str="Multi-byte: 日本語 한국어 العربية עברית 🎉🚀💻",
    )
    utf8_data = utf8_test.encode_cdr2_le()
    utf8_deser, _ = Strings.decode_cdr2_le(utf8_data)

    print("UTF-8 strings preserved:")
    print(f'  Latin-1:    "{utf8_deser.bounded_str}"')
    print(f'  Multi-byte: "{utf8_deser.wide_str}"')

    if utf8_test == utf8_deser:
        print("[OK] UTF-8 encoding preserved correctly")

    # Test long string
    print("\n--- Long String Test ---")
    long_content = ''.join(chr(ord('A') + (i % 26)) for i in range(1000))
    long_str = Strings(unbounded_str=long_content, bounded_str="short", wide_str="also short")
    long_data = long_str.encode_cdr2_le()
    long_deser, _ = Strings.decode_cdr2_le(long_data)

    print(f"Long string length: {len(long_deser.unbounded_str)} chars")
    if long_str == long_deser:
        print("[OK] Long string handled correctly")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())
